src/core/server/saved_objects/import/create_saved_objects.ts

```diff
@@ -83,88 +88,117 @@ export const createSavedObjects = async <T>({
 
   // filter out the 'version' field of each object, if it exists
 
-  const objectsToCreate = filteredObjects.map(({ version, ...object }) => {
-    if (dataSourceId) {
-      // @ts-expect-error
-      if (dataSourceTitle && object.attributes.title) {
-        if (
-          object.type === 'dashboard' ||
-          object.type === 'visualization' ||
-          object.type === 'search'
-        ) {
-          // @ts-expect-error
-          object.attributes.title = object.attributes.title + `_${dataSourceTitle}`;
+  const objectsToCreate = await Promise.all(
+    filteredObjects.map(async ({ version, ...object }) => {
+      if (dataSourceId) {
+        // @ts-expect-error
+        if (dataSourceTitle && object.attributes.title) {
+          if (
+            object.type === 'dashboard' ||
+            object.type === 'visualization' ||
+            object.type === 'search'
+          ) {
+            // @ts-expect-error
+            object.attributes.title = object.attributes.title + `_${dataSourceTitle}`;
+          }
         }
-      }
-
-      if (object.type === 'index-pattern') {
-        object.references = [
-          {
-            id: `${dataSourceId}`,
-            type: 'data-source',
-            name: 'dataSource',
-          },
-        ];
-      }
 
-      if (object.type === 'visualization' || object.type === 'search') {
-        // @ts-expect-error
-        const searchSourceString = object.attributes?.kibanaSavedObjectMeta?.searchSourceJSON;
-        // @ts-expect-error
-        const visStateString = object.attributes?.visState;
+        // Some visualization types will need special modifications, like Vega visualizations
+        if (object.type === 'visualization') {
+          const vegaSpec = extractVegaSpecFromSavedObject(object);
+
+          if (!!vegaSpec && !!dataSourceTitle) {
+            const idComponents = object.id.split('_');
```

Comment:
Do we save the data source id as part of vega vis object id?

Reply:
Yes. The new datasourceid is still part of the vega vis object id and the reference to that datasourceid is also created.